### DISCUSSION TITLE
Improvement: support OAuth2 client credentials instead of ONLY authorization code flow

### DISCUSSION THREAD

**Discussion by nekman** [Thomas55555/husqvarna_automower#510] [Announcements]
Thanks for a great plugin!

Instead of only allowing users to login with the [Authorization code grant](https://oauth.net/2/grant-types/authorization-code/), 
it is possible to login with [client_credentials](https://oauth.net/2/grant-types/client-credentials/) as well.

If you login using the example that you find under "My Applications" / "Your app":
```
curl -X POST -d "grant_type=client_credentials&client_id=<APP_ID>&client_secret=<APP_SECRET>" https://api.authentication.husqvarnagroup.dev/v1/oauth2/token
```

After you have logged in with your client credentials, you will get back a token for your own user (so it's important that the credentials is not shared) where you can control your Automower (if everything is paired and setup).

I think this will make it easier to setup everything for most users since the "My Home Assistant redirect feature" will no longer be needed.

For most users, the `client_credentials` should be enough (you typically use your own secret to login as your self and control your automower). For some edge cases, there might be other APIs and flows where `authorization_grant` is better if you want to allow other users to connect to "your backend" with their token (without sharing their passwords with that backend).

**Comment by Thomas55555** [comment]
Hi @nekman thanks for the idea.
Yes, you are right. A lot of people have to struggle with the Authorization code grant.
I think it was about one year ago. At this time the integration supported log-in with email and the client credentials. It was shortly before login with email/password was shutdown.
At this time I noticed it consumes a lot of time to support to authentication methods. And also at this time with the Application Credential platform it was easier to integrate Authorization code grant in HA. So I decided to use this method, as it seems to be the preferred method.
